semgrep_output_v0.atd

```diff
@@ -119,15 +119,31 @@ type core_match_extra
 type dataflow_trace
   <ocaml attr="deriving show">
   <python decorator="dataclass(frozen=True)"> = {
-  ?taint_source: location list option;
+  ?taint_source: taint_source option;
   (* Intermediate variables which are involved in the dataflow. For taint, this
    * explains how the taint flows from the source to the sink. *)
-  ?intermediate_vars: location list option;
+  ?intermediate_vars: intermediate_var list option;
   (* For now, omitting the taint sink, since it's redundant data except for
    * certain cases that only appear with DeepSemgrep. We can add it in later
    * without breaking consumers. *)
 }
 
+type taint_source
+  <ocaml attr="deriving show">
+  <python decorator="dataclass(frozen=True)"> = {
+  location: location;
+  abstract_content: string;
+}
+
+type intermediate_var
+  <ocaml attr="deriving show">
+  <python decorator="dataclass(frozen=True)"> = {
+  location: location;
+  abstract_content: string;
+  (* This happens to be the same as taint_source for the moment, but Iago has
```

Comment:
Does ATD support recursive data types? Because my extension is mostly for DeepSemgrep (it won't affect plain Semgrep) and it will require recursive data types.

Reply:
It didn't let me use recursive types in an initial version of this datatype. My recollection is that ATD itself does, but atdpy does not. I'm not sure whether it just hasn't been implemented or whether it's due to some limitation of mypy.